File: scripts/fetch-youtube-playlist.js
#!/usr/bin/env node
/**
 * Fetch up to 60 video IDs from a YouTube playlist and output SQL INSERTs
 * for public.theme_songs. Uses YouTube Data API v3.
 *
 * Prerequisites:
 *   1. Create a project in Google Cloud Console.
 *   2. Enable "YouTube Data API v3".
 *   3. Create an API key (Credentials → Create credentials → API key).
 *
 * Usage:
 *   set YOUTUBE_API_KEY=your_key_here
 *   node scripts/fetch-youtube-playlist.js <theme_id> <playlist_id>
 *
 * Example:
 *   node scripts/fetch-youtube-playlist.js a1000000-0000-0000-0000-000000000001 PLrAXerErn2Tl2c3d2_2Fk_6F_EKdF
 *
 * Playlist ID: from a URL like https://www.youtube.com/playlist?list=PLrAXerErn2T...
 *              the id is the part after list=
 *
 * Output: copy the printed SQL and run it in Supabase SQL Editor.
 */

const THEME_ID = process.argv[2]
const PLAYLIST_ID = process.argv[3]
const API_KEY = process.env.YOUTUBE_API_KEY

const MAX_VIDEOS = 60
const PAGE_SIZE = 50

if (!THEME_ID || !PLAYLIST_ID) {
  console.error('Usage: node fetch-youtube-playlist.js <theme_id> <playlist_id>')
  console.error('Example: node fetch-youtube-playlist.js a1000000-0000-0000-0000-000000000001 PLrAXerErn2T...')
  process.exit(1)
}

if (!API_KEY) {
  console.error('Set YOUTUBE_API_KEY environment variable (Google Cloud Console → API key).')
  process.exit(1)
}

async function fetchPlaylistItems(pageToken = '') {
  const params = new URLSearchParams({
    part: 'snippet',
    playlistId: PLAYLIST_ID,
    maxResults: String(PAGE_SIZE),
    key: API_KEY,
  })
  if (pageToken) params.set('pageToken', pageToken)
  const res = await fetch(`https://www.googleapis.com/youtube/v3/playlistItems?${params}`)
  if (!res.ok) {
    const t = await res.text()
    throw new Error(`YouTube API error ${res.status}: ${t}`)
  }
  return res.json()
}

function escapeSql(str) {
  if (str == null) return 'NULL'
  return "'" + String(str).replace(/'/g, "''") + "'"
}

async function main() {
  const videos = []
  let nextPageToken = ''

  while (videos.length < MAX_VIDEOS) {
    const data = await fetchPlaylistItems(nextPageToken)
    const items = data.items || []
    for (const item of items) {
      const vid = item.snippet?.resourceId?.videoId
      const title = item.snippet?.title ?? null
      if (vid) videos.push({ youtube_id: vid, title })
      if (videos.length >= MAX_VIDEOS) break
    }
    nextPageToken = data.nextPageToken || ''
    if (!nextPageToken) break
  }

  if (videos.length < 25) {
    console.error(`Only ${videos.length} videos found. Need at least 25 for bingo.`)
    process.exit(1)
  }

  console.log('-- Paste this into Supabase SQL Editor (theme_songs for theme ' + THEME_ID + ', ' + videos.length + ' songs)')
  console.log('INSERT INTO public.theme_songs (theme_id, youtube_id, title, position) VALUES')

  const values = videos.map((v, i) =>
    `  (${escapeSql(THEME_ID)}, ${escapeSql(v.youtube_id)}, ${escapeSql(v.title)}, ${i})`
  )
  console.log(values.join(',\n') + ';')
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: scripts/fetch_youtube_playlist.py
"""
Fetch up to 60 video IDs from a YouTube playlist and print SQL INSERTs
for public.theme_songs. Uses YouTube Data API v3.

Prerequisites: pip install requests

Usage:
  set YOUTUBE_API_KEY=your_key
  python scripts/fetch_youtube_playlist.py <theme_id> <playlist_id>

Example:
  python scripts/fetch_youtube_playlist.py a1000000-0000-0000-0000-000000000001 PLrAXerErn2T...

Playlist ID: from URL https://www.youtube.com/playlist?list=PLxxx - use the part after list=
"""
import os
import sys

try:
    import requests
except ImportError:
    print("Install requests: pip install requests", file=sys.stderr)
    sys.exit(1)

THEME_ID = sys.argv[1] if len(sys.argv) > 1 else None
PLAYLIST_ID = sys.argv[2] if len(sys.argv) > 2 else None
API_KEY = os.environ.get("YOUTUBE_API_KEY")
MAX_VIDEOS = 60
PAGE_SIZE = 50

if not THEME_ID or not PLAYLIST_ID:
    print("Usage: python fetch_youtube_playlist.py <theme_id> <playlist_id>", file=sys.stderr)
    print("Example: python fetch_youtube_playlist.py a1000000-0000-0000-0000-000000000001 PLrAXerErn2T...", file=sys.stderr)
    sys.exit(1)
if not API_KEY:
    print("Set YOUTUBE_API_KEY environment variable.", file=sys.stderr)
    sys.exit(1)


def escape_sql(s):
    return "NULL" if s is None else "'" + str(s).replace("'", "''") + "'"


videos = []
next_page = None
while len(videos) < MAX_VIDEOS:
    url = "https://www.googleapis.com/youtube/v3/playlistItems"
    params = {"part": "snippet", "playlistId": PLAYLIST_ID, "maxResults": PAGE_SIZE, "key": API_KEY}
    if next_page:
        params["pageToken"] = next_page
    r = requests.get(url, params=params)
    r.raise_for_status()
    data = r.json()
    for item in data.get("items", []):
        vid = item.get("snippet", {}).get("resourceId", {}).get("videoId")
        title = item.get("snippet", {}).get("title")
        if vid:
            videos.append((vid, title))
        if len(videos) >= MAX_VIDEOS:
            break
    next_page = data.get("nextPageToken") or ""
    if not next_page:
        break

if len(videos) < 25:
    print(f"Only {len(videos)} videos. Need at least 25 for bingo.", file=sys.stderr)
    sys.exit(1)

print("-- Paste this into Supabase SQL Editor (theme_songs for theme " + THEME_ID + ", " + str(len(videos)) + " songs)")
print("INSERT INTO public.theme_songs (theme_id, youtube_id, title, position) VALUES")
rows = [f"  ({escape_sql(THEME_ID)}, {escape_sql(v)}, {escape_sql(t)}, {i})" for i, (v, t) in enumerate(videos)]
print(",\n".join(rows) + ";")
